### ISSUE TITLE
Issue With Conflicting Path Parameters

### ISSUE THREAD

**Issue by peterschutt** [starlite-api/starlite#827]
### Discussed in https://github.com/starlite-api/starlite/discussions/825

<div type='discussions-op-text'>

<sup>Originally posted by **stonedprogrammer** November 19, 2022</sup>
Hello,

I didn't want to submit an issue as I wasn't sure if this was an issue on my end and I couldn't find anything related in discussions nor issues. I keep getting the following error:
```
starlite.exceptions.http_exceptions.ImproperlyConfiguredException: 500: Should not use routes with conflicting path parameters
```

with the following controller:
```
class VmcController(Controller):
    path = "/vmc"

    @delete("/{name:str}")
    async def delete_vm(self, name: str) -> None:
        return

    @post("/{id:int}")
    async def restart_vm(self, id: int) -> str:
        return "string"
```

Based on https://starlite-api.github.io/starlite/usage/1-routing/3-controllers/?h=contr, this should be fine as the path + http method combination is unique for both of them, yet each time I get that same error. When I keep the delete route by itself, it works, and vice versa for the post route. Am I doing something incorrectly here?

I have been able to fix it by adding another path level (e.g. changing the delete path to `/delete/{name:str}`, but this doesn't really mesh with the rest of the API routing.</div>

**Comment by peterschutt** [comment]
TL;DR - this can work if the parameters on any path level are the same, i.e. if they were both `{id:int}` or `{name:str}`.

Whether this turns out to be a documentation issue, or something we actually try to resolve via changes to the routing would probably need to come from @Goldziher. However, I look at that example and think we should probably strive to make it work, but I'd need to spend more time trying to understand the feasibility of that.

Path parameters are represented by a single node in the routing trie at any level, e.g., the routing trie for this set of handlers ends up looking like this, prior to the point that the validation raises the error:

```
.
└── /
    └── vmc/
        └── {path parameter}
            ─ DELETE
            ─ OPTIONS
            ─ POST
```
The POST and OPTIONS routes of the 2nd handler _are_ added to the node's "asgi handlers" mapping, but the OPTIONS route of the 2nd handler overwrites that of the first, and nonetheless, immediately thereafter the validation error is raised because the path parameter of the 2nd route is not equal to the path parameter that seeded that trie node (where equality is defined as name ("id", "name"), full name ("name:str", "id:int") and type (`str`, `int`) all being equal).

So if you could define these routes such that they both accepted the same parameter it will work.

For fun, I deleted the validation that raises the error and get the following responses from various requests:

- OPTIONS /vmc/a-name: `allow: OPTIONS, POST` (would be OPTIONS, DELETE)
- OPTIONS /vmc/123: `allow: OPTIONS, POST` (correct)
- DELETE /vmc/a-name: `204 No Content` (correct)
- DELETE /vmc/123: `204 No Content` (arguably correct as the handler with the broader type, `str`, is registered first
- POST /vmc/a-name: `400: value is not a valid integer` (this would be a 404 if parameter type could be considered in routing)
- POST /vmc/123: `400: value is not a valid integer` (this one has me stumped)

For this to work I'd imagine we'd need the trie to end up looking something like this:

```
.
└── Root Node/
    └── vmc/
        ├── {path parameter: int}
        │   ─ OPTIONS
        │   ─ POST
        └── {path parameter: str}
            ─ OPTIONS
            ─ DELETE
```
Where at least in resolving a request path, we'd have a shot at getting it right as long as we tested the parameter value types from most narrow to most broad.

FWIW this does work on starlette:

```python
from starlette.applications import Starlette
from starlette.responses import PlainTextResponse
from starlette.routing import Route


async def homepage(request):
    print(request.path_params)
    return PlainTextResponse("Homepage")

async def about(request):
    print(request.path_params)
    return PlainTextResponse("About")


routes = [
    Route("/{id:int}", endpoint=homepage),
    Route("/{name:str}", endpoint=about),
]

app = Starlette(routes=routes)
```

**Comment by Goldziher** [comment]
So, I don't think we should support ambiguous path params. Frankly, I'd think we should fix the docs raather than try to fix this to work - its going to hurt performance quite significantly.

**Comment by stonedprogrammer** [comment]
Will this issue be added in the roadmap for a future release? I think this is something that limits the ability to use this framework in comparison to other popular ones such as FastAPI and Starlette that do support this type of operation.
